### ISSUE TITLE
Make sure to allow multiple redirect url situation

### ISSUE THREAD

**Issue by weaverryan** [knpuniversity/oauth2-client-bundle#2]
See https://github.com/hwi/HWIOAuthBundle/issues/149 for a discussion. Basically, if we can avoid the headache here, that would be ideal :).

**Comment by okwinza** [comment]
ok, in this case a brief TL;DR:

According to Oauth specs resourceOwners are recommended to implement support for 2 redirect uris(for auth and to connect existing accounts) but they not required to do so.

HWIOauthBundle provides 2 different callback endpoints for each action. It is mostly fine with most mainstream providers but when you face a different one(EveOnline in my case) which only supports 1 callback for everything -- this is where this issue hits you in da face.

Basically, you have to somehow handle both of these actions(auth and connect) on one endpoint.
Im pretty much willing to give you a hand with this if anything.

Hope this will be of any help! :)

**Comment by weaverryan** [comment]
@okwinza Actually, I would *love* your help - I'd like to get this bundle to a useable state soon, but with doing things correctly.

Here are some questions I have:

1) I don't see where you can configure 2 different callback endpoints in HWIOAuthBundle - but probably just because their docs are incomplete. Can you point me to some docs or code?

2) How would you model having 2 potential redirect URIs with https://github.com/thephpleague/oauth2-github? The redirectUri is a constructor argument to the provider. So, would you create one provider for authentication (pointing to your auth callback URI) and a different provider for connecting (pointing to your connect callback URI)?

3) From the OAuth spec point of view, what is the purpose of having 2 redirect URIs? I mean, obviously, one is for authentication and one is for connecting, but what specifically does "connecting" mean? Why not have just one callback URI that handles both? i.e. if you're not authenticated, you'll be authenticated, else just "connect" with the logged in account. I'm playing "devil's advocate" to make sure I understand the purpose. 

4) Some providers don't allow 2 redirect URIs? The redirectUri is often passed as a query parameter to the provider. So if a provider doesn't support 2 redirect URIs, is that basically because they require you to setup an exact redirectUri in the application (and don't allow you to pass anything different as a query param)?

Thanks! Basically, I want to find out the *right* flow for auth and connecting, and make sure the bundle allows you to do that (while keeping things as simple as possible).

**Comment by okwinza** [comment]
@weaverryan ok i've freshen up this mess so let me put it another way:

1) HWIOAuthBundle is bad because it has 2 routes out of the box
(see https://github.com/hwi/HWIOAuthBundle/blob/master/Resources/doc/1-setting_up_the_bundle.md)
"hwi_oauth_redirect" for connecting and "hwi_oauth_login" for login.

I'm not sure why they chose to go this way, maybe to simplify the logic and allow easier integration. At some point later they've decided to handle everything with just 1 endpoint but it didnt happen. 
So i just wanted to warn you to not commit the same mistake.


3) Yep, correct. Having just 1 that handles everything is fine.
4) Yep, some providers force you to specify an exact and only one redirect uri which is using for auth and connect. Passing anything else will result in an error.

So having just 1 redirect URI is perfect in most situations.

**Comment by weaverryan** [comment]
@okwinza do you think that some people *will* want to have multiple URLs? Currently, I was envisioning that if you want a Facebook OAuth provider, this bundle gives you exactly 1: with *one* redirectUri. But, perhaps we need the ability for people to create 2 (or 3 or 4) Facebook providers with different redirectUri's.

**Comment by bshaffer** [comment]
> According to Oauth specs resourceOwners are recommended to implement support for 2 redirect uris(for auth and to connect existing accounts) but they not required to do so.

@okwinza I did a quick search through the [openid spec](http://openid.net/specs/openid-connect-core-1_0.html) and [oauth2 spec](https://tools.ietf.org/html/rfc6749), and do not see any mention of this. Could you provide a reference to the spec you are referring to?

@weaverryan the [OAuth2 spec](https://tools.ietf.org/html/rfc6749#section-3.1.2.2) has this to say:

>    The authorization server MAY allow the client to register multiple redirection endpoints.

You'll notice the wording there makes it optional, so in the case of Eve Online, they are still abiding by the spec.

The client library should be able to configure more than one redirect URI, but in some cases may not be able to, depending on the oauth server's configuration. This is one of the reasons people complain about the (lack of) interoperability of OAuth2... some oauth2 servers can end up being implemented quite differently. See the [Interopability](https://tools.ietf.org/html/rfc6749#section-1.8) section for perhaps my favorite quote of the whole spec:

> as a rich and highly extensible
   framework with many optional components, on its own, this
   specification is likely to produce a wide range of non-interoperable
   implementations.

I hope this clarifies some things.

**Comment by bshaffer** [comment]
Also, I would love to hear more about this *auth redirect* vs *existing account redirect*. Is this a convention used for separating concerns? The *auth redirect* utilizes openID connect to create an account and the *existing account redirect* simply grabs a refresh token / access token?

**Comment by weaverryan** [comment]
I've updated the bundle to support multiple redirect URLs. Put simply, if you want, you can create two "Facebook" provider objects, each with their own `redirect_uri`. So, you can be as flexible as you want :).

**Comment by okwinza** [comment]
@weaverryan :+1:
